### DISCUSSION TITLE
Luarocks/busted setup on macOS

### DISCUSSION THREAD

**Discussion by fredrikaverpil** [HiPhish/rainbow-delimiters.nvim#110] [Q&A] [answered]
Hi @HiPhish, following [this reddit thread](https://www.reddit.com/r/neovim/comments/1cicjno/comment/l2bc1jz/), I'm attempting to just run the test suite locally on macOS, but I feel something is a little strange with my local setup, and I was thinking maybe you can spot what I'm doing wrong.

I'm using pkgx to install lua and luarocks. I'm using a [pkgx](https://docs.pkgx.sh/) with the [ability to set tooling on a per-project basis](https://docs.pkgx.sh/using-dev/dev), but I think you can just achieve the similar result by doing this:

```sh
brew install pkgxdev/made/pkgx
pkgx install luajit.org
pkgx install luarocks.org
```

Then I install busted:

```sh
# install and use luajit
luarocks --lua-dir=$HOME/.pkgx/luajit.org/v2.1.0 --lua-version=5.1 install busted

# make busted available on path
PATH=$PATH:$HOME/.luarocks/bin

# set LUA_PATH
luarocks --lua-dir=$HOME/.pkgx/luajit.org/v2.1.0 --local path
```

Finally, I run the tests:

```sh
make unit-test

✱
0 successes / 0 failures / 1 error / 0 pending : 0.000312 seconds

Error → ./test/unit/stack_spec.lua @ 1
suite ./test/unit/stack_spec.lua
./test/unit/stack_spec.lua:1: module 'rainbow-delimiters.stack' not found:
        no field package.preload['rainbow-delimiters.stack']
        no file './src/rainbow-delimiters/stack.lua'
        no file './src/rainbow-delimiters/stack/rainbow-delimiters/stack.lua'
        no file './src/rainbow-delimiters/stack/init.lua'
        no file '/Users/fredrik/.pkgx/luarocks.org/v3.11.0/tbin/../share/lua/5.1/rainbow-delimiters/stack.lua'
        no file '/Users/fredrik/.pkgx/luajit.org/v2.1.0/share/luajit-2.1/rainbow-delimiters/stack.lua'
        no file '/Users/fredrik/.luarocks/share/lua/5.1/rainbow-delimiters/stack.lua'
        no file '/Users/fredrik/.luarocks/share/lua/5.1/rainbow-delimiters/stack/init.lua'
        no file '/Users/fredrik/.pkgx/luarocks.org/v3.11.0/share/lua/5.1/rainbow-delimiters/stack.lua'
        no file '/Users/fredrik/.pkgx/luarocks.org/v3.11.0/share/lua/5.1/rainbow-delimiters/stack/init.lua'
        no file './csrc/rainbow-delimiters/stack.so'
        no file './csrc/rainbow-delimiters/stack/rainbow-delimiters/stack.so'
        no file '/usr/local/lib/lua/5.1/rainbow-delimiters/stack.so'
        no file '/usr/local/lib/lua/5.1/loadall.so'
        no file './rainbow-delimiters/stack.so'
        no file '/Users/fredrik/.luarocks/lib/lua/5.1/rainbow-delimiters/stack.so'
        no file '/Users/fredrik/.pkgx/luarocks.org/v3.11.0/lib/lua/5.1/rainbow-delimiters/stack.so'
        no file './csrc/rainbow-delimiters.so'
        no file './csrc/rainbow-delimiters/rainbow-delimiters.so'
        no file '/usr/local/lib/lua/5.1/rainbow-delimiters.so'
        no file '/usr/local/lib/lua/5.1/loadall.so'
        no file './rainbow-delimiters.so'
        no file '/Users/fredrik/.luarocks/lib/lua/5.1/rainbow-delimiters.so'
        no file '/Users/fredrik/.pkgx/luarocks.org/v3.11.0/lib/lua/5.1/rainbow-delimiters.so'
E5113: Error while calling lua chunk: [NULL]
make: *** [unit-test] Error 1
```

This is where I guess I have done something wrong...?

If I just copy the project into a `src` folder, I can run the tests just fine:

```sh
cp -r lua src

make unit-test
●●●●●●●●●●
10 successes / 0 failures / 0 errors / 0 pending : 0.002297 seconds
```

I'm not super well adversed with lua/luarocks/busted, and so I'm not sure what is dictating these search paths. Any clue as to what I need to change?

**Comment by HiPhish** [comment]
You have changed the `PATH` environment variable which is not necessary. The file `test/busted` acts a wrapper around the real busted executable, it will evaluate `luarocks path --lua-version 5.1 --bin` to adjust the `PATH` for the duration of the test.

Something in your setup is messing up the search paths where Neovim is looking for Lua modules. I will be out of town for the duration of the next week, so please don't be surprised if I don't respond until then. You can ping me here if I don't reply for a longer time.

**Reply by fredrikaverpil** [reply]
> You have changed the PATH environment variable which is not necessary.

Your wrapper calls `busted`, but that's not on my `$PATH` when installing it this way. I don't have `~/.luarocks/bin` on my $PATH, which I am simply making clear I have there.

Yes, indeed, something is messing up the search paths.

**Reply by HiPhish** [reply]
> Your wrapper calls `busted`, but that's not on my `$PATH` when installing it this way.

And it does not have to be on your `PATH`. My wrapper calls `luarocks path ...` first before calling `busted`. This will adjust the `PATH` variable for the duration of the test only. Lua is tricky because users are expected to have multiple versions of Lua installed on their system and later versions are not backwards-compatible with prior versions (i.e. Lua 5.2 is not a strict superset of Lua 5.1). This means I cannot assume that the `busted` in your `~/.luarocks/bin` is the correct one. By letting LuaRocks adjust the `PATH` temporarily for Lua 5.1 I can be certain that the correct busted is used.

**Reply by fredrikaverpil** [reply]
Yikes, okay I see. You're right, I don't have to set the `$PATH`. 👍

**Reply by fredrikaverpil** [reply, marked as answer]
I'm not sure what I'm doing but by adding this to the `busted` wrapper, the tests are now found on my machine without resorting to renaming any directories, when I run `make unit-test`:

```bash
export LUA_PATH="./lua/?.lua;;"
export LUA_CPATH="./lua/?.so;;"
```

**Comment by HiPhish** [comment]
Oh wait, what directory are you running the tests from? You need to be *inside* the plugin directory, neither above nor below it. Otherwise all the relative paths are wrong. There is a `./src` directory if your search path, but I have no idea where this is coming from.

Another thing you can try is unsetting the `LUA_PATH` and `LUA_CPATH` environment variables before running tests. As long a the `luarocks` binary is on your `PATH` that's all that matters.

**Reply by fredrikaverpil** [reply]
> what directory are you running the tests from?

I'm running `make unit-test` from the root of the git repo.

> There is a ./src directory if your search path, but I have no idea where this is coming from.

Yes, exactly ... I've got no clue either 😄 

> Another thing you can try is unsetting the LUA_PATH and LUA_CPATH environment variables

I'll try what you're mentioning and follow up here later. 👍 Thanks!
